File: src/main/java/com/salesforce/dynamodbv2/mt/mappers/MtAmazonDynamoDb.java
package com.salesforce.dynamodbv2.mt.mappers;

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.model.Identity;
import com.amazonaws.services.dynamodbv2.model.OperationType;
import com.amazonaws.services.dynamodbv2.model.Record;
import com.amazonaws.services.dynamodbv2.model.ScanResult;
import com.amazonaws.services.dynamodbv2.model.StreamRecord;
import java.util.List;

/**
 * This interface (including all contained interfaces and methods) is
 * experimental. It is subject to breaking changes. Use at your own risk.
 */
public interface MtAmazonDynamoDb extends AmazonDynamoDB {

    class MtRecord extends Record {

        private static final long serialVersionUID = -6099434068333437314L;

        private String context;
        private String tableName;

        public String getContext() {
            return context;
        }

        public MtRecord withContext(String context) {
            this.context = context;
            return this;
        }

        public String getTableName() {
            return tableName;
        }

        public MtRecord withTableName(String tableName) {
            this.tableName = tableName;
            return this;
        }

        @Override
        public MtRecord withAwsRegion(String awsRegion) {
            setAwsRegion(awsRegion);
            return this;
        }

        @Override
        public MtRecord withDynamodb(StreamRecord dynamodb) {
            setDynamodb(dynamodb);
            return this;
        }

        @Override
        public MtRecord withEventID(String eventId) {
            setEventID(eventId);
            return this;
        }

        @Override
        public MtRecord withEventName(OperationType eventName) {
            setEventName(eventName);
            return this;
        }

        @Override
        public MtRecord withEventName(String eventName) {
            setEventName(eventName);
            return this;
        }

        @Override
        public MtRecord withEventSource(String eventSource) {
            setEventSource(eventSource);
            return this;
        }

        @Override
        public MtRecord withEventVersion(String eventVersion) {
            setEventVersion(eventVersion);
            return this;
        }

        @Override
        public MtRecord withUserIdentity(Identity userIdentity) {
            setUserIdentity(userIdentity);
            return this;
        }

        @Override
        public String toString() {
            return "MtRecord{"
                + "context='" + context + '\''
                + ", tableName='" + tableName + '\''
                + ", recordFields=" + super.toString()
                + '}';
        }
    }

    /**
     * Wrap a scan result from a multi tenant scan, to associate tenant information per scan result item.
     **/
    class MtScanResult extends ScanResult {
        private final List<String> tenants;
        private final List<String> virtualTables;

        public MtScanResult(ScanResult scanResult, List<String> tenants, List<String> virtualTables) {
            super();
            this.withLastEvaluatedKey(scanResult.getLastEvaluatedKey());
            this.withConsumedCapacity(scanResult.getConsumedCapacity());
            this.withCount(scanResult.getCount());
            this.withScannedCount(scanResult.getScannedCount());
            this.setSdkHttpMetadata(scanResult.getSdkHttpMetadata());
            this.setSdkResponseMetadata(scanResult.getSdkResponseMetadata());
            this.tenants = tenants;
            this.virtualTables = virtualTables;
        }

        public List<String> getTenants() {
            return tenants;
        }

        public List<String> getVirtualTables() {
            return virtualTables;
        }
    }

}

File: src/main/kotlin/com/salesforce/dynamodbv2/mt/mappers/MtAmazonDynamoDb.kt
/* Copyright (c) 2019, Salesforce.com, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause.
 * For full license text, see LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause.
 */
package com.salesforce.dynamodbv2.mt.mappers

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
import com.amazonaws.services.dynamodbv2.model.Identity
import com.amazonaws.services.dynamodbv2.model.OperationType
import com.amazonaws.services.dynamodbv2.model.Record
import com.amazonaws.services.dynamodbv2.model.StreamRecord

/**
 * This interface (including all contained interfaces and methods) is
 * experimental. It is subject to breaking changes. Use at your own risk.
 */
interface MtAmazonDynamoDb : AmazonDynamoDB {
    class MtRecord : Record() {

        var context: String? = null
            private set
        var tableName: String? = null
            private set

        fun withContext(context: String): MtRecord {
            this.context = context
            return this
        }

        fun withTableName(tableName: String): MtRecord {
            this.tableName = tableName
            return this
        }

        override fun withAwsRegion(awsRegion: String): MtRecord {
            setAwsRegion(awsRegion)
            return this
        }

        override fun withDynamodb(dynamodb: StreamRecord): MtRecord {
            setDynamodb(dynamodb)
            return this
        }

        override fun withEventID(eventId: String): MtRecord {
            eventID = eventId
            return this
        }

        override fun withEventName(eventName: OperationType): MtRecord {
            setEventName(eventName)
            return this
        }

        override fun withEventName(eventName: String): MtRecord {
            setEventName(eventName)
            return this
        }

        override fun withEventSource(eventSource: String): MtRecord {
            setEventSource(eventSource)
            return this
        }

        override fun withEventVersion(eventVersion: String): MtRecord {
            setEventVersion(eventVersion)
            return this
        }

        override fun withUserIdentity(userIdentity: Identity): MtRecord {
            setUserIdentity(userIdentity)
            return this
        }

        override fun toString(): String {
            return ("MtRecord{" +
                    "context='" + context + '\''.toString() +
                    ", tableName='" + tableName + '\''.toString() +
                    ", recordFields=" + super.toString() +
                    '}'.toString())
        }

        companion object {

            private val serialVersionUID = -6099434068333437314L
        }
    }

    data class TenantTable(val virtualTableName: String, val tenantName: String)
}
